org.eclipse.xtext/src/org/eclipse/xtext/generator/URIBasedFileSystemAccess.java

```diff
@@ -0,0 +1,244 @@
+	}
+
+	public interface BeforeRead {
+		InputStream beforeRead(URI changed, InputStream in);
+	}
+
+	private URIConverter converter;
+	private URI baseDir;
+	private boolean generateTraces = false;
+	private IEncodingProvider encodingProvider = new IEncodingProvider.Runtime();
+	private TraceRegionSerializer traceRegionSerializer;
+	private TraceFileNameProvider traceFileNameProvider;
+	private URIBasedFileSystemAccess.BeforeDelete beforeDelete = (uri) -> true;
+	private URIBasedFileSystemAccess.BeforeWrite beforeWrite = (uri, string, stream) -> stream;
+	private URIBasedFileSystemAccess.BeforeRead beforeRead = (uri, stream) -> stream;
+
+	@Override
+	public void setPostProcessor(IFilePostProcessor filePostProcessor) {
+		super.setPostProcessor(filePostProcessor);
+	}
+
+	@Override
+	public URI getURI(String path, String outputConfiguration) {
+		String outlet = getPathes().get(outputConfiguration);
+		if (outlet == null) {
+			throw new IllegalArgumentException(
+					("A slot with name \'" + outputConfiguration) + "\' has not been configured.");
+		}
+		URI uri = URI.createFileURI(outlet + File.separator + path);
+		return baseDir != null ? uri.resolve(baseDir) : uri;
+	}
+
+	public String getEncoding(URI uri) {
+		return encodingProvider.getEncoding(uri);
+	}
+
+	@Override
+	public void generateFile(String fileName, String outputCfgName, CharSequence contents) {
+		try {
+			URI uri = getURI(fileName, outputCfgName);
+			if (!getOutputConfig(outputCfgName).isOverrideExistingResources()
+					&& converter.exists(uri, CollectionLiterals.emptyMap())) {
+				return;
+			}
+			String encoding = getEncoding(uri);
+			CharSequence postProcessed = postProcess(fileName, outputCfgName, contents, encoding);
+			generateTrace(fileName, outputCfgName, postProcessed);
+			generateFile(fileName, outputCfgName,
+					new ByteArrayInputStream(postProcessed.toString().getBytes(encoding)));
+		} catch (Throwable e) {
+			throw Exceptions.sneakyThrow(e);
+		}
+	}
+
+	protected void generateTrace(String generatedFile, String outputConfigName, CharSequence contents) {
+		try {
+			if (isGenerateTraces() && contents instanceof ITraceRegionProvider) {
+				try {
+					ByteArrayOutputStream out = new ByteArrayOutputStream();
+					traceRegionSerializer.writeTraceRegionTo(((ITraceRegionProvider) contents).getTraceRegion(), out);
+					generateFile(traceFileNameProvider.getTraceFromJava(generatedFile), outputConfigName,
+							new ByteArrayInputStream(out.toByteArray()));
+				} catch (Throwable t) {
+					if (t instanceof TraceNotFoundException) {
+						// do not re-throw
+					} else {
+						throw Exceptions.sneakyThrow(t);
+					}
+				}
+			}
+		} catch (Throwable t) {
+			throw Exceptions.sneakyThrow(t);
+		}
+	}
+
+	@Override
+	public void generateFile(String fileName, String outputCfgName, InputStream content) throws RuntimeIOException {
+		try {
+			URI uri = getURI(fileName, outputCfgName);
+			try (OutputStream out = converter.createOutputStream(uri)) {
```

Comment:
Not sure if `out` should better be buffered if it isn't already.

Reply:
Not sure, too. As this is refactoring only I tend to not change behaviour.